scipy/stats/_morestats.py

```diff
@@ -2739,6 +2740,10 @@ def ansari(x, y, alternative='two-sided'):
     if alternative not in {'two-sided', 'greater', 'less'}:
         raise ValueError("'alternative' must be 'two-sided',"
                          " 'greater', or 'less'.")
+
+    if not hasattr(_abw_state, 'a'):
+        _abw_state.a = _ABW()
```

Comment:
Could you add a comment in the code about how this works and why the name of the attribute is `a` (or `s` in some tests)? Same with the MWU state. Thanks!

Reply:
See https://docs.python.org/3/library/threading.html#thread-local-data. This pattern is the same everwhere - avoid mutable global state. Note that the `.a` name is irrelevant. I think @andfoy has chosen one-letter names to keep line lengths short; it could also have been named `.val` or something like that.

The more detailed explanation here: `_ABW` is a class whose methods mutate attributes of the instance. If you have an `_ABW` instance that is global to the module, then there is nothing preventing (in the absence of the GIL) parallel calls to `_ABW` methods from multiple threads. That clearly is going to go wrong (e.g., thread 1 is in the middle of something, and thread 2 overwrites `self.freqs`). So the solution is to make the stateful thing thread-local.

Reply:
I added these comments, and included them in the PR split off from this one (gh-22175).